File: src/murmur/bed.py
"""Background-bed acquisition + cache (spec 03-04 §2.2/§2.3).

The bed audio is a curated manifest of yt-dlp-resolvable refs committed to the
repo (``assets/bed_sources.txt``), pulled to a per-user cache during first-run
loading. At runtime the engine plays only the local cached files — no network,
no resolve latency (that is the dead-air spec 04 exists to hide). This module
owns the manifest reader, the local ``CachedBedSource`` the engine consumes, and
the loading-time pull; ``make bed-refresh`` re-pulls after the manifest changes.
"""

from __future__ import annotations

import asyncio
import contextlib
import hashlib
from collections.abc import Awaitable, Callable
from pathlib import Path

from . import paths
from .logging_setup import get_log

_log = get_log("bed")

# The committed manifest and the per-user cache (outside the repo, survives
# clones) — spec 03-04 §2.3 "cache location — decided". The cache is rebuildable
# state, so it lives under murmur's cache root (spec 05 §2.3 path governance);
# $MURMUR_HOME relocates it.
DEFAULT_MANIFEST = Path(__file__).resolve().parent.parent.parent / "assets" / "bed_sources.txt"
DEFAULT_CACHE_DIR = paths.cache_root() / "bed"

# Partial / hidden files are not playable tracks.
_SKIP_SUFFIXES = {".part", ".ytdl", ".tmp"}


def read_manifest(path: Path) -> list[str]:
    """Refs from the manifest, one per line; ``#`` comments and blanks skipped.
    A missing manifest is simply empty (degrade to no bed)."""
    if not path.exists():
        return []
    refs: list[str] = []
    for raw in path.read_text(encoding="utf-8").splitlines():
        line = raw.strip()
        if line and not line.startswith("#"):
            refs.append(line)
    return refs


def _cache_key(ref: str) -> str:
    """A stable filename stem for a ref, so a warm cache is recognised across
    runs regardless of the ref's characters."""
    return hashlib.sha256(ref.encode("utf-8")).hexdigest()[:16]


def _cached_files(cache_dir: Path) -> list[Path]:
    if not cache_dir.is_dir():
        return []
    return sorted(
        p
        for p in cache_dir.iterdir()
        if p.is_file() and not p.name.startswith(".") and p.suffix not in _SKIP_SUFFIXES
    )


class CachedBedSource:
    """The runtime ``BedSource`` (spec 03-04 §2.2): lists the cached local bed
    files in a stable order. No network — resolving happened at loading time."""

    def __init__(self, cache_dir: Path) -> None:
        self._cache_dir = cache_dir

    def tracks(self) -> list[Path]:
        return _cached_files(self._cache_dir)


async def ytdlp_download(ref: str, dest_base: Path, *, ytdlp: str = "yt-dlp") -> None:
    """Pull one ref's best audio to ``<dest_base>.<ext>`` via yt-dlp (reusing the
    03-01 acquisition binary). Loading-time only — never on the audio path."""
    proc = await asyncio.create_subprocess_exec(
        ytdlp,
        "-f",
        "bestaudio/best",
        "-o",
        f"{dest_base}.%(ext)s",
        ref,
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
    )
    _stdout, stderr = await proc.communicate()
    if proc.returncode != 0:
        detail = stderr.decode(errors="replace").strip()
        raise RuntimeError(f"yt-dlp failed ({proc.returncode}): {detail}")


async def pull_bed(
    *,
    manifest: Path,
    cache_dir: Path,
    download: Callable[[str, Path], Awaitable[None]],
    log: Callable[[str], None] = _log.event,
) -> int:
    """First-run pull (spec 03-04 §2.3): resolve each manifest ref into the cache
    via ``download``, skipping already-cached refs and continuing past a failing
    one (a dead ref never aborts the pull). Returns the number of cached tracks
    afterward — 0 (offline / empty manifest) degrades cleanly to no bed."""
    refs = read_manifest(manifest)
    if not refs:
        return len(_cached_files(cache_dir))
    cache_dir.mkdir(parents=True, exist_ok=True)
    failures = 0
    for ref in refs:
        key = _cache_key(ref)
        if any(cache_dir.glob(f"{key}.*")):
            continue  # warm cache — skip, no network
        log(f"bed: pulling {ref}")
        try:
            await download(ref, cache_dir / key)
        except Exception as exc:  # a dead ref must not abort the pull
            failures += 1
            # The raw yt-dlp error (e.g. a 403 on one source) is noise to the
            # user: the pull degrades cleanly, so keep the detail in the debug
            # log and surface only a calm summary below.
            _log.debug(f"bed: pull failed for {ref!r}: {exc}")
    cached = len(_cached_files(cache_dir))
    if failures:
        log(
            f"bed: {failures} source(s) unavailable, skipped "
            f"({cached}/{len(refs)} ready)"
        )
    return cached


def main() -> None:
    """``make bed-refresh``: (re-)pull the committed manifest into the cache."""
    count = asyncio.run(
        pull_bed(
            manifest=DEFAULT_MANIFEST,
            cache_dir=DEFAULT_CACHE_DIR,
            download=ytdlp_download,
            log=print,
        )
    )
    print(f"bed: {count} track(s) cached in {DEFAULT_CACHE_DIR}")


if __name__ == "__main__":  # pragma: no cover
    with contextlib.suppress(KeyboardInterrupt):
        main()

File: src/bed.ts
// Background-bed acquisition + cache (spec 03-04 §2.2/§2.3): a curated manifest
// of yt-dlp refs committed to the repo, pulled to the per-user cache at loading
// time. At runtime the engine plays only the local cached files — no network on
// the audio path. Cache layout and key match the Python implementation, so an
// already-warm cache is reused as-is across the port.

import { execFile } from 'node:child_process'
import { createHash } from 'node:crypto'
import { existsSync, mkdirSync, readdirSync, readFileSync, renameSync, writeFileSync } from 'node:fs'
import { mkdir, readFile } from 'node:fs/promises'
import { basename, join } from 'node:path'
import { fileURLToPath } from 'node:url'

import { z } from 'zod'

import type { BedPosition, BedSource } from './contracts.ts'
import { cacheRoot } from './paths.ts'

// fileURLToPath, not URL.pathname: pathname keeps %-escapes (a checkout path
// with a space would silently read as an empty manifest).
export const DEFAULT_MANIFEST = fileURLToPath(new URL('../assets/bed_sources.txt', import.meta.url))

export function defaultBedCacheDir(env: NodeJS.ProcessEnv = process.env): string {
  return join(cacheRoot(env), 'bed')
}

// Partial / hidden files are not playable tracks.
const SKIP_SUFFIXES = ['.part', '.ytdl', '.tmp']

// Refs one per line; `#` comments and blanks skipped. Missing manifest = empty.
export async function readManifest(path: string): Promise<string[]> {
  const text = await readFile(path, 'utf8').catch(() => '')
  return text
    .split('\n')
    .map((line) => line.trim())
    .filter((line) => line !== '' && !line.startsWith('#'))
}

// A stable filename stem per ref — sha256(ref)[:16], the Python convention, so
// a warm cache is recognised across runs and across the language port.
export function cacheKey(ref: string): string {
  return createHash('sha256').update(ref, 'utf8').digest('hex').slice(0, 16)
}

function cachedFiles(cacheDir: string): string[] {
  if (!existsSync(cacheDir)) return []
  return readdirSync(cacheDir)
    .filter((name) => !name.startsWith('.') && !SKIP_SUFFIXES.some((s) => name.endsWith(s)))
    .sort()
    .map((name) => join(cacheDir, name))
}

// Where the bed left off, beside the tracks it points into (spec 03-04 resume):
// wiping the cache wipes the position with it, which is the right semantic. The
// leading dot keeps it out of cachedFiles' track listing. `track` is a basename,
// so the position survives a relocated MURMUR_HOME.
const POSITION_FILE = '.position.json'

const BedPositionSchema = z.object({ track: z.string().min(1), offsetS: z.number().nonnegative() })

// A missing, damaged, or hand-mangled file is simply "no position" — the bed
// starts from the top, never a boot failure.
export function readBedPosition(cacheDir: string): BedPosition | null {
  let raw: string
  try {
    raw = readFileSync(join(cacheDir, POSITION_FILE), 'utf8')
  } catch {
    return null
  }
  try {
    const checked = BedPositionSchema.safeParse(JSON.parse(raw))
    return checked.success ? checked.data : null
  } catch {
    return null
  }
}

// Temp-file + rename (the settings discipline) so a reader never sees a torn file.
export function writeBedPosition(cacheDir: string, position: BedPosition): void {
  mkdirSync(cacheDir, { recursive: true })
  const path = join(cacheDir, POSITION_FILE)
  writeFileSync(`${path}.tmp`, `${JSON.stringify(position)}\n`, 'utf8')
  renameSync(`${path}.tmp`, path)
}

// Map a saved basename back onto the cached track list. undefined = no resume
// (nothing saved, or the track has left the cache) — start from the top.
export function resumeFrom(tracks: string[], saved: BedPosition | null): BedPosition | undefined {
  if (saved === null) return undefined
  const track = tracks.find((t) => basename(t) === saved.track)
  return track === undefined ? undefined : { track, offsetS: saved.offsetS }
}

// The tail we never aim a seek into: landing inside the final crossfade plays a
// blink of audio and rolls over — start earlier (or from the top) instead.
const RESUME_TAIL_S = 10

// The boot-time start decision (spec 03-04 resume). A saved position whose
// offset fits inside the track's real duration wins. Everything else — first
// boot, a vanished track, a stale offset past the end — lands on a RANDOM track
// at a RANDOM in-bounds offset, so no two fresh boots open on the same bars.
// `durationOf` returning null (no ffprobe, unreadable file) degrades the pick
// to offset 0 and keeps a saved offset as-is: the engine's miss handling owns
// whatever the probe could not rule out.
export async function initialBedPosition(
  tracks: string[],
  saved: BedPosition | null,
  durationOf: (track: string) => Promise<number | null>,
  random: () => number = Math.random,
): Promise<BedPosition | undefined> {
  const resumed = resumeFrom(tracks, saved)
  if (resumed !== undefined) {
    const duration = await durationOf(resumed.track)
    if (duration === null || resumed.offsetS < Math.max(0, duration - RESUME_TAIL_S)) return resumed
    // stale offset: fall through to a fresh random start
  }
  if (tracks.length === 0) return undefined
  const track = tracks[Math.floor(random() * tracks.length) % tracks.length]!
  const duration = await durationOf(track)
  if (duration === null || duration <= RESUME_TAIL_S) return { track, offsetS: 0 }
  return { track, offsetS: random() * (duration - RESUME_TAIL_S) }
}

// The runtime BedSource (spec 03-04 §2.2): local cached files, stable order.
export class CachedBedSource implements BedSource {
  private cacheDir: string

  constructor(cacheDir: string = defaultBedCacheDir()) {
    this.cacheDir = cacheDir
  }

  tracks(): string[] {
    return cachedFiles(this.cacheDir)
  }
}

export type BedDownload = (ref: string, destBase: string) => Promise<void>

export type PullBedOptions = {
  manifest: string
  cacheDir: string
  download: BedDownload
  log?: (message: string) => void
}

// First-run pull (spec 03-04 §2.3): resolve each manifest ref into the cache,
// skipping warm refs and continuing past failures — a dead ref never aborts the
// pull. Returns the cached-track count; 0 degrades cleanly to no bed.
export async function pullBed({ manifest, cacheDir, download, log }: PullBedOptions): Promise<number> {
  const refs = await readManifest(manifest)
  if (refs.length === 0) return cachedFiles(cacheDir).length
  await mkdir(cacheDir, { recursive: true })
  let failures = 0
  for (const ref of refs) {
    const key = cacheKey(ref)
    if (cachedFiles(cacheDir).some((path) => path.includes(`/${key}.`))) continue
    log?.(`bed: pulling ${ref}`)
    try {
      await download(ref, join(cacheDir, key))
    } catch {
      failures += 1 // detail is noise to the user; the pull degrades cleanly
    }
  }
  const cached = cachedFiles(cacheDir).length
  if (failures > 0) log?.(`bed: ${failures} source(s) unavailable, skipped (${cached}/${refs.length} ready)`)
  return cached
}

// Pull one ref's best audio to `<destBase>.<ext>` via yt-dlp (the 03-01
// acquisition binary, reused). Loading-time only — never on the audio path.
export function ytdlpDownload(ref: string, destBase: string, ytdlpCmd = 'yt-dlp'): Promise<void> {
  return new Promise((resolve, reject) => {
    execFile(
      ytdlpCmd,
      ['-f', 'bestaudio/best', '-o', `${destBase}.%(ext)s`, ref],
      { timeout: 300_000 },
      (err, _stdout, stderr) => {
        if (err) reject(new Error(`yt-dlp failed: ${String(stderr).trim() || err.message}`))
        else resolve()
      },
    )
  })
}
